Emit new `AgentMetadataEvent`

vitorenesduarte opened the pull request:
Part of https://github.com/gravitational/cloud/issues/3550.
Depends on https://github.com/gravitational/cloud/pull/3554.

This PR adds a new `UpstreamInventoryAgentMetadata` that is sent from a Teleport agent to an auth server.
Once received, it is transformed into an `AgentMetadataEvent` and sent to PreHog.

Commit:
Emit new `AgentMetadataEvent`

Part of https://github.com/gravitational/cloud/issues/3550.

This commits adds a new `UpstreamInventoryAgentMetadata` that is sent
from a Teleport agent to an auth server.
Once received, it is transformed into an `AgentMetadataEvent` and sent
to PreHog.

Most `UpstreamInventoryAgentMetadata` fields are intentionally kept as
empty in this PR. Follow up PRs will be opened with the mechanisms
required to fill them (as described in #21337).

Commit:
server_id -> host_id

Commit:
merge master

Commit:
compute OS and host architecture

Commit:
Compute OS version and container runtime

Commit:
Close stream if agent receives an agent metadata message

Commit:
cmd -> exec ; file -> read

Commit:
implement fetchOSVersion for linux

Commit:
Remove unused import

Commit:
Add note about `agentMetadataCh`

Commit:
merge master

Commit:
Allow commands with args

Commit:
Remove parseFun abstraction

Commit:
fetch glibc version

Commit:
fetch container orchestrator

Commit:
Fix lint

Commit:
Merge branch 'master' into vitor/agent-metadata-event

Commit:
Fix TODO

Commit:
Add note about glibc version

Commit:
cmd -> command

Commit:
Merge branch 'master' into vitor/agent-metadata-event

espadolini commented on lib/inventory/metadata.go:
```diff
@@ -0,0 +1,233 @@
+// fetchHostArchitecture computes the host architecture using the arch
+// command-line utility.
+func (c *fetchConfig) fetchHostArchitecture() string {
+	command := "arch"
+	out, err := c.exec(command)
```
Can `arch` ever return a value that's different than what we can infer from `runtime.GOARCH`? On macOS, at least, it seems to output `i386` when called from within a Rosetta-enabled binary.

vitorenesduarte replied:
Yes. This was the motivation behind using `arch` instead of `runtime.GOARCH`.

Commit:
fetch cloud environment

espadolini commented on lib/inventory/metadata.go:
```diff
@@ -0,0 +1,314 @@
+}
+
+// fetchContainerRuntime returns "docker" if the file "/.dockerenv" exists.
+func (c *fetchConfig) fetchContainerRuntime() string {
```
Why not use something like https://pkg.go.dev/github.com/genuinetools/bpfd@v0.0.1/proc#GetContainerRuntime to get more complete info?

vitorenesduarte replied:
For this initial phase, the RFD proposes that we only try to track Docker. Happy to add this now if we feel that is valuable.

(`gopsutil` also detects a lot of runtimes: https://github.com/shirou/gopsutil/blob/v3.23.1/internal/common/common_linux.go#L130-L278)

timothyb89 replied:
`genuinetools/bpfd` won't detect modern docker versions since it relies on `/proc/foo/cgroup` being populated, which is no longer the case. `gopsutil` looks like it should work, though.

espadolini commented on lib/inventory/metadata.go:
```diff
@@ -0,0 +1,314 @@
+		return false
+	}
+
+	return c.httpReqSuccess(req)
```
Will the metadata version endpoint work even if the metadata service is disabled?

vitorenesduarte replied:
Good point. It won't with something like [this](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instance-metadata-limiting-access), but we probably shouldn't try to workaround that.

espadolini replied:
We should probably set a fairly tight deadline on those HTTP requests so that they don't block the rest of the data too much.

vitorenesduarte replied:
Added a 5 minute timeout in 15eb1a7525bf63d49824c02bcbc6251838eaa89b. Please LMK if it should be smaller.

espadolini commented on lib/inventory/metadata.go:
```diff
@@ -0,0 +1,314 @@
+// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
+func (c *fetchConfig) awsHttpGetSuccess() bool {
+	url := "http://169.254.169.254/latest/meta-data/"
+	req, err := http.NewRequest(http.MethodGet, url, nil)
```
Here and below.

```suggestion
	req, err := http.NewRequestWithContext(c.ctx, http.MethodGet, url, nil)
```

vitorenesduarte replied:
Thanks! Done in 883117cd82b37a276dd6ec1e957fb1cee9743ec9.

espadolini commented on lib/inventory/metadata.go:
```diff
@@ -0,0 +1,314 @@
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
```
If we want to keep a string for this rather than use an enum:

```suggestion
// Service names for UpstreamInventoryAgentMetadata
const (
```

espadolini commented on lib/inventory/inventory.go:
```diff
@@ -119,6 +124,14 @@ type downstreamHandle struct {
 	senderC      chan DownstreamSender
 	closeContext context.Context
 	cancel       context.CancelFunc
+	// agentMetadataCh is the buffered channel (size 1) where the agent metadata
+	// will be sent to at most once (once it is calculated). For this reason, if
+	// the agent metadata is taken out of the channel and we fail to send it
+	// upstream, then the metadata will never be sent upstream as there is no
+	// retry mechanism. However, since the agent metadata is sent after the
+	// hello exchange succeeds, it is unlikely that the sending of the metadata
+	// will fail.
+	agentMetadataCh chan proto.UpstreamInventoryAgentMetadata
```
The way this seems to be coded is that we do one attempt to send the metadata per lifetime of the `TeleportProcess`, we should be sending the metadata once per connection to an auth server instead.

vitorenesduarte replied:
Yes, this is intentional.

The initial proposal in the RFD would send 1 `UpstreamInventoryAgentMetadata` per `UpstreamInventoryHello`.
However, this complicates the implementation a bit without any clear benefit (as most agent metadata won't change throughout the lifetime of the agent).

espadolini replied:
I'm not sure I like that, it's gonna be far too easy to lose track of the data like that, and in the current model the auth server would be forced to persist the data somewhere if we wanted to keep it available for the user.

vitorenesduarte replied:
The easiest way to send 1 `UpstreamInventoryAgentMetadata` per `UpstreamInventoryHello` would be to spawn the goroutine for each new stream: fe4f2790be2db87c1aea1271a037ef72072f91bf

WDYT? Could this spawn many unnecessary goroutines?

---

Alternatively, (as proposed initially in the RFD: https://github.com/gravitational/teleport/commit/7737d46912a990f37c10e2778c21a3dc1a52af02#diff-dbf83bde2c5bcf10c7f0b59b4ed0f2720ff90e8d0ca33f1f4ad2702344299f01) we could spawn a single that:
- computes the metadata just once, and
- sends the metadata through a channel when it is requested

Commit:
fetch install methods

Commit:
GLibCVersion -> GlibcVersion

vitorenesduarte commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -2055,12 +2056,41 @@ message UpstreamInventoryHello {
+  // HostArchitecture advertises the instance host architecture (e.g. "x86_64" or "arm64").
+  string HostArchitecture = 6;
+  // GLibCVersion advertises the instance glibc version of linux instances (e.g. "2.35").
+  string GLibCVersion = 7;
```
Done in 9c2af8ac686ee49ecf1eb2d2d6e85e10c40271fd.

Commit:
Use `http.NewRequestWithContext`

Commit:
Add missing comment

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Commit:
Merge branch 'master' into vitor/agent-metadata-event

Commit:
GLibCVersion -> GlibcVersion

Commit:
Fix lint

Commit:
Fix helm unit tests

Commit:
Add missing comments

Commit:
5 second timeout on http requests

espadolini commented on lib/inventory/metadata.go:
```diff
@@ -0,0 +1,366 @@
+	}
+	if c.httpDo == nil {
+		c.httpDo = func(req *http.Request) (*http.Response, error) {
+			client := &http.Client{}
```
We seem to use this throughout the codebase.
```suggestion
			client := defaults.HTTPClient()
```

Commit:
Spawn gorountine that fetches metadata on each new stream

Commit:
Use `defaults.HTTPClient()`

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Commit:
Add missing import

Commit:
Handle error

Commit:
Merge branch 'master' into vitor/agent-metadata-event

espadolini commented on lib/inventory/inventory.go:
```diff
@@ -162,6 +162,14 @@ func (h *downstreamHandle) tryRun(fn DownstreamCreateFunc, hello proto.UpstreamI
 func (h *downstreamHandle) handleStream(stream client.DownstreamInventoryControlStream, upstreamHello proto.UpstreamInventoryHello) error {
+	agentMetadataCh := make(chan proto.UpstreamInventoryAgentMetadata, 1)
+	go func() {
+		agentMetadataCh <- fetchAgentMetadata(&fetchConfig{
```
We can cache the result of this as part of `downstreamHandle`, no need to do this once per stream.

Something like
```go
type downstreamHandle struct {
	...

	agentMetadataDone chan struct {} // to be initialized with no buffer
	agentMetadata proto.UpstreamInventoryAgentMetadata
}

// in NewDownstreamHandle
go func() {
	h.agentMetadata = fetchAgentMetadata(...)
	close(h.agentMetadataDone)
}

// in handleStream
...
agentMetadataDone := h.agentMetadataDone
select  {
...
case <-agentMetadataDone:
	agentMetadataDone = nil
	if err := stream.Send(h.closeContext, h.agentMetadata); err != nil {
```

vitorenesduarte replied:
Thanks!! Cool trick 😮 

Done in fd03367813142cb3f485288f507c2e1f7dee0545.

Commit:
Revert "Spawn gorountine that fetches metadata on each new stream"

This reverts commit fe4f2790be2db87c1aea1271a037ef72072f91bf.

Commit:
Send agent metadata to auth server once per stream

Commit:
Improve note about agentMetadataDone

wadells commented:
In short, LGTM*

* unless you anticipate that this data will ever be rendered/persisted/processed on the auth server, in which case the `proto.UpstreamInventoryAgentMetadata` handler needs sanity checking logic.

From a security perspective, the current regex validation on agents is good UX, but irrelevant since it could be bypassed by an malicious client (e.g. a custom teleport version or an edited replay attack).  We want to sanitize on the auth server (if we ever think this data will be rendered on in the auth server). But most importantly, we want to sanitize incoming census data in infrastructure we control -- which is out of scope for this PR since that needs to happen in prehog.  Any logic in the Teleport repo is the untrusted client for prehog.

Commit:
Don't process command output & file content on the agent

Since agents cannot be trusted, regex validation and sanitization should
happen in PreHog anyways. So this commit removes such logic in favor of
moving it to PreHog.

vitorenesduarte commented:
> From a security perspective, the current regex validation on agents is good UX, but irrelevant since it could be bypassed by an malicious client (e.g. a custom teleport version or an edited replay attack). We want to sanitize on the auth server (if we ever think this data will be rendered on in the auth server). But most importantly, we want to sanitize incoming census data in infrastructure we control -- which is out of scope for this PR since that needs to happen in prehog. Any logic in the Teleport repo is the untrusted client for prehog.

This is a good point, thanks!

Since agents cannot be trusted and sanitization needs to happen in PreHog anyways, e3b18a45e395b1e4c4a4c3122e1bf8e66075f239 removes the parsing/validation/sanitization in favor of moving such logic to PreHog.

This also helps in improving the parsing & validation code as this project moves along since we can deploy a new version of PreHog much faster than agents are upgraded.

I'll open a follow up to https://github.com/gravitational/cloud/pull/3554, updating PreHog with this suggestion.

Commit:
Merge branch 'master' into vitor/agent-metadata-event

Commit:
Fix lint

fspmarshall commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -2055,12 +2056,43 @@ message UpstreamInventoryHello {
+// UpstreamInventoryAgentMetadata is the message sent up the inventory control stream containing
+// metadata about the instance.
+message UpstreamInventoryAgentMetadata {
+  // Version advertises the teleport version of the instance.
+  string Version = 1;
+  // HostID advertises the ID of the instance.
+  string HostID = 2;
+  // Services advertises the currently live services of the instance that represent
+  // access protocols ("node", "kube", "app", "db" and "windows_desktop").
+  repeated string Services = 3;
```
Nit: The `Hello` message already includes this data, and the hello is cached by the auth server for the lifetime of the stream.  Rather than double-sending the same data, I think it would be preferable to only send new data, and have the auth server construct the audit event from a combination of `Hello` and `AgentMetadata` params.

Ex:

```golang
func (c *Controller) handleAgentMetadata(handle *upstreamHandle, msg proto.UpstreamInventoryAgentMetadata) error {
    if err := c.auth.AnonymizeAndSubmit(&services.AgentMetadataEvent{
    	Version:               handle.Hello().Version,
 	HostId:                handle.Hello().ServerID,
 	Services:              handle.Hello().Services,
 	Os:                    msg.OS,
 	OsVersionInfo:         msg.OSVersionInfo,
 	HostArchitectureInfo:  msg.HostArchitectureInfo,
 	...
     }); err != nil {
 	log.Debugf("Unable to submit agent metadata: %v", err)
     }
}
```

Minor difference I know, but per-agent network load is a very serious scalability concern in large clusters.  Anything we can do to minimize load is a win.

vitorenesduarte replied:
Thanks for the suggestion! Done in fa89947abf1917e6051d6107ee3ae94bc8bc9174.

Commit:
Trim space

Commit:
Merge branch 'master' into vitor/agent-metadata-event

fspmarshall commented on lib/inventory/inventory.go:
```diff
@@ -87,12 +87,19 @@ type DownstreamSender interface {
-		pingHandlers: make(map[uint64]DownstreamPingHandler),
-		closeContext: ctx,
-		cancel:       cancel,
+		senderC:           make(chan DownstreamSender),
+		pingHandlers:      make(map[uint64]DownstreamPingHandler),
+		closeContext:      ctx,
+		cancel:            cancel,
+		agentMetadataDone: make(chan struct{}),
 	}
 	go handle.run(fn, hello)
+	go func() {
+		handle.agentMetadata = fetchAgentMetadata(&fetchConfig{ctx: ctx, hello: hello})
+		// Signal that the agentMetadata has been calculated by closing the
+		// agentMetadataDone channel.
+		close(handle.agentMetadataDone)
+	}()
```
This is more of a stylistic comment, so take it or leave it...

I think this code might be more maintainable and easier to reason about if we tried to keep it as loosely-coupled as possible.  Rather than integrating the metadata fetching/sending into the core `DownstreamHandle` logic, I'd like to propose an alternative that should achieve the same effect (example assumes we also get rid of the need to re-send hello fields):


```golang
func AutoEmitMetadata(handle DownstreamHandle) {
    metadata := fetchAgentMetadata(handle.CloseContext())
    for {
        select {
        case sender := <-handle.Sender(): // wait for stream to be opened
            sender.Send(handle.CloseContext(), metadata) // send metadata
            select {
            case <-sender.Done(): // block for the duration of the stream
            case <-handle.CloseContext.Done():
                return
            }
        case <-handle.CloseContext.Done():
            return
        }
    }
}
```

The above will have the effect of ensuring that metadata is sent once each time a stream is established, but it has the nice effect of being very loosely coupled with the internals of the handle.

This is already how control-stream based heartbeats work (except they send periodically for the life of the stream), and I think it's a good pattern to stick to if possible.  I think `lib/inventory` is complex enough, that any time we can break out components to be separate that's a win.

vitorenesduarte replied:
Done in cbdefedc8040014de974167c680ca23dd49c9052.

fspmarshall reviewed (commented):
Basic idea LGTM.  Added some notes on how to better conform to the inventory control stream's design patterns.

Also, one general suggestion: if possible, could we break out the metadata fetching into a separate namespace (e.g. `lib/inventory/metadata`)?

The `lib/inventory` package can be complex/daunting, so I'd like to limit its contents to only be the "glue code" as much as possible.  Just message routing/handling, with all the business logic kept elsewhere.

Commit:
Move handling of `AgentMetadataEvent` to `handleControlStream`

Commit:
Use cached hello message

Commit:
Move metadata files to lib/inventory/metadata

vitorenesduarte commented on lib/auth/auth.go:
```diff
@@ -3062,15 +3062,33 @@ func (a *Server) MakeLocalInventoryControlStream(opts ...client.ICSPipeOption) c
+				}); err != nil {
+					log.Debugf("Unable to submit agent metadata: %v", err)
+				}
+			default:
```
Thanks for the review! I moved this to `handleControlStream` in ab0f33ff6227a6b4c2dd8e33c2ed551cb9ce4ecd.

Commit:
make sending of agent metadata more self contained

Commit:
Minimize diff

Commit:
Merge branch 'master' into vitor/agent-metadata-event

fspmarshall commented on lib/inventory/metadata/metadata.go:
```diff
@@ -0,0 +1,334 @@
+	}
+}
+
+// FetchServices computes the list of access protocols enabled at the agent from
+// the list of system roles present in the hello message.
+func FetchServices(systemRoles []types.SystemRole) []string {
+	var services []string
+	for _, svc := range systemRoles {
+		switch svc {
+		case types.RoleNode:
+			services = append(services, nodeService)
+		case types.RoleKube:
+			services = append(services, kubeService)
+		case types.RoleApp:
+			services = append(services, appService)
+		case types.RoleDatabase:
+			services = append(services, dbService)
+		case types.RoleWindowsDesktop:
+			services = append(services, windowsDesktopService)
+		}
+	}
+	return services
+}
```
Nit: Probably better to avoid needing to update `inventory/metadata` any time we add a new system role.  I'd suggest adding a method like `IsAccessService` to `types.SystemRole`.  There is existing precedence for using the methods of `types.SystemRole` to distinguish what kind of role it is (e.g. `role.IsServerRole()`, which is like what you are doing here, but also includes auth/proxy).

Similarly, I'd either use the canonical names, or apply a standardized transform (e.g. `strings.ToLower(...)`), rather than hard-coding alternative names.

That way we can end up w/ something like this:

```golang
func NormalizeAndFilterServices(systemRoles []types.SystemdRole) []string {
    var services []string
    for _, svc := range systemRoles {
        if !svc.IsAccessService() {
            // skip service roles not directly associated with an access protocol (e.g. auth/proxy)
            continue
        }
        services = append(services, strings.ToLower(string(svc)))
    }
    return services
}
```

vitorenesduarte replied:
Thanks! There was no motivation (IIRC) in having *only* the system roles that correspond to an access service, and so 64ec45d7d21c2fae8ae05c24524586c1664bc0d8 changes this in order to send all system roles to PreHog.

fspmarshall commented on lib/inventory/metadata/metadata.go:
```diff
@@ -0,0 +1,334 @@
+// in tracking.
+func FetchAgentMetadata(c *AgentMetadataFetchConfig) proto.UpstreamInventoryAgentMetadata {
+	c.setDefaults()
+	return proto.UpstreamInventoryAgentMetadata{
+		OS:                    c.fetchOS(),
+		OSVersionInfo:         c.fetchOSVersionInfo(),
+		HostArchitectureInfo:  c.fetchHostArchitectureInfo(),
+		GlibcVersionInfo:      c.fetchGlibcVersionInfo(),
+		InstallMethods:        c.fetchInstallMethods(),
+		ContainerRuntime:      c.fetchContainerRuntime(),
+		ContainerOrchestrator: c.fetchContainerOrchestrator(),
+		CloudEnvironment:      c.fetchCloudEnvironment(),
+	}
```
Nit: I bet a lot of this info will be useful in contexts outside of just the metadata message/event.  Low-priority ask, but if you have some free time, it'd be nifty to have these individual fetch operations be a public API that we could leverage elsewhere in the codebase without doing a full fetch of all parameters.

Commit:
Send all system roles to PreHog

Commit:
merge master

Commit:
Remove unused import

Commit:
Add parsing of command output / file content back

vitorenesduarte commented:
As discussed in https://github.com/gravitational/cloud/pull/3626, 00cfbfecd791d434abef27d8abdb164712afc196 adds the parsing back to the agent code.

Commit:
Usage reporter refactor

Commit:
Merge branch 'master' into vitor/agent-metadata-event

Commit:
Usage reporter refactor

Commit:
Add missing handling of inventory agent metadata msg

Commit:
Fix ICS usage reporter

Commit:
Improve comments

Commit:
Add cached `metadata.Get*` methods

Commit:
Use systemctl status instead of is-active

Commit:
Merge branch 'master' into vitor/agent-metadata-event

Commit:
Add `Metadata` struct

Commit:
return pointer in `FetchMetadata`

Commit:
Pass context to `GetMetadata`

Commit:
metadataFetchConfig -> fetchConfig

Commit:
GetMetadata -> Get

Commit:
Add note about `Get` result

Commit:
Ensure install methods are non-nil

Commit:
Exit `metadata.Get` if context is closed

Commit:
Replace sync.Once with atomic.Bool.Swap

wadells commented:
Removing myself as a reviewer, as the important part of my concerns is now handled in https://github.com/gravitational/cloud/pull/3664

Commit:
Merge branch 'master' into vitor/agent-metadata-event

Commit:
Initialize channel

Commit:
Fix lint

Commit:
Fix lint

Commit:
Merge branch 'master' into vitor/agent-metadata-event

espadolini commented on lib/inventory/metadata/metadata.go:
```diff
@@ -0,0 +1,322 @@
+// dockerfileInstallMethod returns true if the instance was installed using our
+// Dockerfile.
+func (c *fetchConfig) dockerfileInstallMethod() bool {
+	return c.getenv("TELEPORT_INSTALL_METHOD_DOCKERFILE") == "true"
```
```suggestion
	return c.getenv("TELEPORT_INSTALL_METHOD_DOCKERFILE") != ""
```
or maybe something fancier with `utils.ParseBool`, allowing an explicit `0` or `false` while letting everything else count as true?

vitorenesduarte replied:
Done in bff342e63298429a393f729455a6ae624dc763bc.

espadolini commented on lib/inventory/metadata/metadata.go:
```diff
@@ -0,0 +1,322 @@
+
+// systemctlInstallMethod returns true if the instance is running using systemctl.
+func (c *fetchConfig) systemctlInstallMethod() bool {
+	out, err := c.exec("systemctl", "status", "teleport.service")
+	if err != nil {
+		return false
+	}
+
+	return strings.Contains(out, "active (running)")
```
We should probably replace this with something that uses the systemd dbus API in the future.

vitorenesduarte replied:
Thanks! Added this as a possible improvement in https://github.com/gravitational/cloud/issues/3550#issue-1591930330.

espadolini commented on lib/inventory/metadata/metadata_linux.go:
```diff
@@ -0,0 +1,71 @@
+// fetchGlibcVersion parses the output of 'ldd --version' and returns e.g.
+// "2.31".
+func (c *fetchConfig) fetchGlibcVersion() string {
+	out, err := c.exec("ldd", "--version")
+	if err != nil {
+		return ""
+	}
+
+	lines := strings.Split(strings.TrimSpace(out), "\n")
+	if len(lines) == 0 {
+		return ""
+	}
+
+	// The glibc version is the last word on the first line.
+	parts := strings.Fields(lines[0])
+	return parts[len(parts)-1]
```
This will fail quite badly when running in alpine with gcompat. Can we call `gnu_get_libc_version` via cgo instead? I'm pretty sure that teleport doesn't actually compile with non-glibc anyway. cc @zmb3

vitorenesduarte replied:
Also added this idea as a possible improvement in https://github.com/gravitational/cloud/issues/3550#issue-1591930330.

espadolini replied:
Launching `ldd` will also fail in the new distroless images, they don't have `ldd`.

espadolini replied:
That also applies to `arch`, btw.

vitorenesduarte replied:
Added this to https://github.com/gravitational/cloud/issues/3550#issue-1591930330 as well.

vitorenesduarte replied:
After discussing with @espadolini, we're now:
- using `gnu_get_libc_version` (ddcc8c5e2d23e7bd4c911bf0c6c2c1d6cd2d0722) as this seems to be a much more reliable approach
- using `GOARCH` (b694ca9a196b43af1d893df048d68e57fd9b88dc) as there doesn't seem to be a likely situation in which its output would differ from `arch`

espadolini commented on lib/inventory/metadata/metadata_linux.go:
```diff
@@ -0,0 +1,71 @@
+	var name string
+	var versionID string
+	for _, line := range strings.Split(strings.TrimSpace(out), "\n") {
+		parts := strings.Split(line, "=")
```
Use `strings.Cut` instead.

vitorenesduarte replied:
Done in 08c431bbd60afad02a079c1048d46a44df344777.

espadolini commented on lib/inventory/metadata/metadata_linux.go:
```diff
@@ -0,0 +1,71 @@
+		}
+
+		switch parts[0] {
+		case "NAME":
```
ID is intended to be machine-parsable, we're gonna have much better luck with that one instead.

vitorenesduarte replied:
Done in 3fbe859053accc172503f5438c2972e8aa1d20ee. This deviates from what was suggested in the RFD but I think it makes sense to do so.

Commit:
Make `metadata.Get` return an error instead of bool

vitorenesduarte commented on lib/inventory/metadata/get.go:
```diff
@@ -0,0 +1,58 @@
+// modified.
+func Get(ctx context.Context) (*Metadata, bool) {
+	if !fetched.Swap(true) {
+		// Spawn a goroutine responsible for fetching the metadata if we're the
+		// first Get caller.
```
Using `atomic.Bool.Swap` allows each caller to have its own cancellation context: the gorountine will be spawn at most once and callers unblock as soon as their context is cancelled.

If we use `sync.Once` only the cancellation context of the first caller will be respected.

vitorenesduarte commented on lib/inventory/metadata/get.go:
```diff
@@ -0,0 +1,58 @@
+		// Spawn a goroutine responsible for fetching the metadata if we're the
+		// first Get caller.
+		go func() {
+			defaultFetcher := &fetchConfig{}
+			defaultFetcher.setDefaults()
+			metadata = defaultFetcher.fetch()
```
I think it's fine to use `context.Background()` here as the `Get` caller context is respected below.

vitorenesduarte commented on lib/inventory/metadata/get.go:
```diff
@@ -0,0 +1,58 @@
+	case <-metadataReady:
+		return metadata, true
+	case <-ctx.Done():
+		return nil, false
```
Done in 0eb945aaf526e8c52d5f034c27b3bc5e24341288.

Commit:
Allow multiple true/false values for env vars

Commit:
Use `strings.Cut`

Commit:
Use /etc/os-release ID instead of NAME

vitorenesduarte commented:
@espadolini I've addressed all comments. Could you please take another look?

Commit:
merge master

Commit:
Improve `autoEmitMetadata`

Commit:
Use `gnu_get_libc_version`

Commit:
Ubuntu -> ubuntu

Commit:
Use GOARCH

Commit:
merge master

espadolini commented on lib/inventory/metadata/metadata_linux.go:
```diff
@@ -0,0 +1,62 @@
+	for _, line := range strings.Split(strings.TrimSpace(out), "\n") {
+		key, value, found := strings.Cut(line, "=")
+		if !found {
+			return ""
```
> Blank lines are permitted and ignored.


```suggestion
			continue
```

Commit:
gofmt

espadolini commented on lib/inventory/metadata/metadata_linux.go:
```diff
@@ -0,0 +1,62 @@
+
+		switch key {
+		case "ID":
+			id = value
```
> Variable assignment values must be enclosed in double or single quotes if they include spaces, semicolons or other special characters outside of A–Z, a–z, 0–9. (Assignments that do not include these special characters may be enclosed in quotes too, but this is optional.)
```suggestion
			id = strings.Trim(value, `"`)
```

espadolini commented on lib/inventory/metadata/metadata_linux.go:
```diff
@@ -0,0 +1,62 @@
+		return ""
+	}
+
+	var id string
+	var versionID string
```
> [ID] If not set, a default of "ID=linux" may be used.

> [VERSION_ID] This field is optional.

```suggestion
	id := "linux"
	versionID := "(unknown)"
```

espadolini reviewed (commented):
Quoting https://www.freedesktop.org/software/systemd/man/os-release.html

Commit:
Move import C up

Commit:
Variables may include quotes

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Commit:
Default values for ID and VERSION_ID

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Commit:
Blank lines are permitted

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

espadolini commented on lib/usagereporter/teleport/types.go:
```diff
@@ -334,6 +334,29 @@ func (u *SFTPEvent) Anonymize(a utils.Anonymizer) prehogv1.SubmitEventRequest {
+		Event: &prehogv1.SubmitEventRequest_AgentMetadataEvent{
+			AgentMetadataEvent: &prehogv1.AgentMetadataEvent{
+				Version:               u.Version,
+				HostId:                u.HostId,
```
This might not be just a UUID.
```suggestion
				HostId:                a.AnonymizeString(u.HostId),
```

vitorenesduarte replied:
Thanks for catching this! 20a7c22321721e5fb989efe826119b70dae2efba

Commit:
Anonymize host id

Commit:
Merge branch 'master' into vitor/agent-metadata-event

Commit:
Merge branch 'master' into vitor/agent-metadata-event
